.builder/actions/pkcs11_test_setup.py

```diff
@@ -0,0 +1,159 @@
+import re
+
+
+class Pkcs11TestSetup(Builder.Action):
+    """
+    Set up this machine for running the PKCS#11 tests.
+    If SoftHSM2 is not installed, the tests are skipped.
+
+    This action should be run in the 'pre_build_steps' or 'build_steps' stage.
+    """
+
+    def run(self, env):
+        self.env = env
+
+        softhsm_lib = self._find_softhsm_lib()
+        if softhsm_lib is None:
+            print("WARNING: libsofthsm2.so not found. PKCS#11 tests are disabled")
+            return
+
+        # set cmake flag so PKCS#11 tests are enabled
+        env.project.config['cmake_args'].append('-DENABLE_PKCS11_TESTS=ON')
+
+        # put SoftHSM config file and token directory under the build dir.
+        softhsm2_dir = os.path.join(env.build_dir, 'softhsm2')
+        conf_path = os.path.join(softhsm2_dir, 'softhsm2.conf')
+        token_dir = os.path.join(softhsm2_dir, 'tokens')
+        env.shell.mkdir(token_dir)
+        self._setenv('SOFTHSM2_CONF', conf_path)
+        with open(conf_path, 'w') as conf_file:
+            conf_file.write(f"directories.tokendir = {token_dir}\n")
+
+        # create a token
+        self._exec_softhsm2_util(
+            '--init-token',
+            '--free', # use any free slot
+            '--label', 'my-test-token',
+            '--pin', '0000',
+            '--so-pin', '0000')
+
+        # we need to figure out which slot the new token is in because:
+        # 1) old versions of softhsm2-util make you pass --slot <number>
+        #    (instead of accepting --token <name> like newer versions)
+        # 2) newer versions of softhsm2-util reassign new tokens to crazy
+        #    slot numbers (instead of simply using 0 like older versions)
+        slot = self._get_token_slots()[0]
+
+        # add private key to token
+        resources_dir = '../../tests/resources'
+        this_dir = os.path.dirname(__file__)
+        resources_dir = os.path.realpath(os.path.join(this_dir, resources_dir))
+        self._exec_softhsm2_util(
+            '--import', os.path.join(resources_dir, 'unittests.p8'),
+            '--slot', str(slot),
+            '--label', 'my-test-key',
+            '--id', 'BEEFCAFE', # ID is hex
+            '--pin', '0000')
+
+        # for logging's sake, print the new state of things
+        self._exec_softhsm2_util('--show-slots', '--pin', '0000')
+
+        # set env vars for tests
+        self._setenv('TEST_PKCS11_LIB', softhsm_lib)
+        self._setenv('TEST_PKCS11_TOKEN_LABEL', 'my-test-token')
+        self._setenv('TEST_PKCS11_PIN', '0000')
+        self._setenv('TEST_PKCS11_PKEY_LABEL', 'my-test-key')
+        self._setenv('TEST_PKCS11_CERT_FILE', os.path.join(resources_dir, 'unittests.crt'))
+        self._setenv('TEST_PKCS11_CA_FILE', os.path.join(resources_dir, 'unittests.crt'))
+
+
+    def _find_softhsm_lib(self):
+        """Return path to SoftHSM2 shared lib, or None if not found"""
+        for lib_dir in ['lib64', 'lib']: # search lib64 before lib
```

Comment:
This won't work for LD_LIBRARY_PATH or an ld.conf script, which some package managers will screw with.

something like `ldconfig -p` is what you'd normally use to find where the system thinks libraries are